hoomd/write/hdf5.py

```diff
@@ -0,0 +1,295 @@
+
+    def __getstate__(self):
+        state = copy.copy(self.__dict__)
+        del state["_fh"]
+        state["_attached_"] = False
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__ = state
+
+
+class HDF5Logger(_InternalCustomWriter):
+    """An HDF5 logger backend.
+
+    This class handles scalar and array data storing them in HDF5 resizable
+    datasets.
+
+    Note:
+        This class requires that ``h5py`` be installed.
+
+    Important:
+        The HDF5 file can be used for other data storage; however, the
+        "hoomd-data" key is reserved for use by this class. An exception will be
+        thrown if this requirement is not met.
+
+    Warning:
+        This class cannot handle string, strings, or object loggables.
+
+    Args:
+        trigger (hoomd.trigger.trigger_like): The trigger to determine when to
+            run the HDF5 backend.
+        filename (str): The filename of the HDF5 file to write to.
+        logger (hoomd.logging.Logger): The logger instance to use for querying
+            log data.
+        mode (`str`, optional): The mode to open the file in. Available values
+            are "w", "x" and "w-", "a", and "r+". Defaults to "a". See the
+            h5py_ documentation for more details).
+
+    .. _h5py:
+        https://docs.h5py.org/en/stable/high/file.html#opening-creating-files
+
+    Attributes:
+        cls.accepted_categories (hoomd.logging.LoggerCategories): The enum value
+            for all accepted categories for `HDF5Logger` instances which is all
+            categories other than "string", "strings", and "object" (see
+            `hoomd.logging.LoggerCategories`).
+        trigger (hoomd.trigger.trigger_like): The trigger to determine when to
+            run the HDF5 backend.
+        filename (str): The filename of the HDF5 file written to.
+        logger (hoomd.logging.Logger): The logger instance used for querying
+            log data.
+        mode (str): The mode the file was opened in.
+
+    .. rubric:: Example
+
+    .. invisible-code-block: python
+
+        import hoomd
+
+        try:
+            import h5py
+            h5py_not_available = False
+        except ImportError:
+            h5py_not_available = True
+
+
+        simulation = hoomd.util.make_example_simulation()
+        h5_filename = tmp_path / "simulation_log.h5"
+
+    .. skip: start if(h5py_not_available)
+
+    .. code-block:: python
+
+        logger = hoomd.logging.Logger(
+            hoomd.write.HDF5Logger.accepted_categories)
+        h5_writer = hoomd.write.HDF5Logger(
+            trigger=10_000,
+            filename=h5_filename,
+            logger=logger)
+        simulation.operations += h5_writer
+    """
+    _internal_class = _HDF5LoggerInternal
+    _wrap_methods = ("flush",)
+
+    def write(self, timestep=None):
+        """Write out data to the HDF5 file.
+
+        Writes out a frame from the composed logger.
+        """
```

Comment:
If I understand correctly, users can call `write` to add frames to the log that are not triggered by `trigger`.
We need to:
1) Add an example snippet.
2) Test `write` in the unit test.
and 3) Explain clearly that calling `write` will add an additional frame even if not triggered.

One issue with `write` is that the fields requiring any of the the flags `ROTATIONAL_KINETIC_ENERGY`, `PRESSURE_TENSOR`, or `EXTERNAL_FIELD_VIRIAL` may be NaN. Those quantities may or may not be available depending on whether the last timestep executed had those flags set by some other means.